lisa/runner.py

```diff
@@ -257,77 +287,69 @@ def _start_loop(self) -> None:
-            while task_manager.wait_worker() or remaining_runners:
-                assert task_manager.has_idle_worker()
-                has_idle_worker = True
-
-                # runners shouldn't mark them done, until all task completed. It
-                # can be checked by test results status or other signals.
-                assert remaining_runners, (
-                    f"no remaining runners, but there are running tasks. "
-                    f"{task_manager._futures}"
+
+            # submit tasks until idle workers are available
+            while task_manager.has_idle_worker():
+                self._submit_tasks(remaining_runners, task_manager)
+
+                # remove completed runners
+                remaining_runners = [
+                    runner for runner in remaining_runners if not runner.is_done
+                ]
+                self._log.debug(
+                    f"remaining runners {[x.id for x in remaining_runners]} "
+                    f"Idle workers : {len(task_manager._futures)}"
                 )
 
-                for runner in remaining_runners:
-                    while not runner.is_done:
-                        # fetch a task and submit
-                        task = runner.fetch_task()
-                        if task:
-                            # the message may be too long to display.
-                            task_message = str(task)
-                            task_message = (
-                                task_message
-                                if len(task_message) < 200
-                                else f"{task_message[:200]}..."
-                            )
-                            self._log.debug(
-                                f"fetched task from {runner.id}: '{task_message}'"
-                            )
-                            task_manager.submit_task(task)
-                        else:
-                            # current runner may not be done, but it doesn't
-                            # have task temporarily. The root runner can start
-                            # tasks from next runner.
-                            break
-                        if not task_manager.has_idle_worker():
-                            has_idle_worker = False
-                            break
-                    if runner.is_done:
-                        # remove fully completed runner.
-                        runner.close()
-                        remaining_runners.remove(runner)
+                if task_manager.has_idle_worker():
+                    if has_more_runner:
+                        # add new runner upto max concurrency if idle workers
+                        # are available
+                        try:
+                            while True:
+                                remaining_runners.append(next(runner_iterator))
+                                if len(remaining_runners) > self._max_concurrency:
```

Comment:
This condition can be moved to while condition.
`while len(remaining_runners) < self._max_concurrency:`

Reply:
We want to atleast add one runner if there are idle workers. It's like do...while loop which runs atleast one time rather than a while loop which can run zero times

Reply:
The logic will run at first time like a do...while. The remaining runners should be empty at first time.

Reply:
But what about the time when len(runner) == self._max_concurrency but no runners have any tasks available?